src/net/udp/frame.rs

```diff
@@ -70,30 +18,33 @@ pub trait UdpCodec {
 pub struct UdpFramed<C> {
     socket: UdpSocket,
     codec: C,
-    rd: Vec<u8>,
-    wr: Vec<u8>,
+    rd: BytesMut,
+    wr: BytesMut,
     out_addr: SocketAddr,
     flushed: bool,
 }
 
-impl<C: UdpCodec> Stream for UdpFramed<C> {
-    type Item = C::In;
+impl<C: Decoder> Stream for UdpFramed<C> {
+    type Item = (C::Item, SocketAddr);
     type Error = C::Error;
 
-    fn poll(&mut self) -> Poll<Option<C::In>, C::Error> {
-        let (n, addr) = try_nb!(self.socket.recv_from(&mut self.rd));
+    fn poll(&mut self) -> Poll<Option<(Self::Item)>, Self::Error> {
+        let mut buf = vec![0; 64 * 1024]; // FIXME reuse self.rd somehow
+        let (n, addr) = try_nb!(self.socket.recv_from(&mut buf));
```

Comment:
Please help me with this part

Reply:
Here is what the decoder should be:

```rust
impl<C: Decoder> Stream for UdpFramed<C> {
    type Item = (C::Item, SocketAddr);
    type Error = C::Error;

    fn poll(&mut self) -> Poll<Option<(Self::Item)>, Self::Error> {
        trace!("udp/frame.rs UdpFramed/Stream/poll");
        self.rd.reserve(INITIAL_RD_CAPACITY);

        let addr = unsafe {
            // Read into the buffer without having to initialize the memory.
            let (n, addr) = try_nb!(self.socket.recv_from(self.rd.bytes_mut()));
            self.rd.advance_mut(n);
            trace!("udp/frame.rs UdpFramed/Stream/ received {} bytes, decoding", n);
            addr
        };

        let frame = self.codec.decode(&mut self.rd)?;
        self.rd.clear();
        let result = frame.map(|frame| (frame, addr)); // frame -> (frame, addr)
        trace!("udp/frame.rs UdpFramed/Stream/ frame decoded from buffer");
        Ok(Async::Ready(result))
    }
}
```

The issue is that the decode trait expects the passed in buf to *only* contain the data that was read.

This unsafe snippet allows reading into uninitialized memory to avoid the cost of constantly zeroing out memory.

Reply:
Thank you so much!